### PR TITLE
autoload MultiContainerInterface

### LINKED ISSUE #888: get_class not working for MultiContainerInterfaces
We would like to create MultiContainerInterface extensions for the Frank Lab. We have gotten them working using the pattern described in the [Extensions tutorial](https://pynwb.readthedocs.io/en/latest/tutorials/general/extensions.html#creating-and-using-a-custom-multicontainerinterface), which utilizes nested NWBGroupSpecs. However, we would like to just call [get_class](https://pynwb.readthedocs.io/en/latest/pynwb.html#pynwb.get_class) to get a working MultiContainerInterface class instead of having to write the Python classes ourselves.

We were able to get nested groups working, thanks to @bendichter et al. in (https://github.com/NeurodataWithoutBorders/pynwb/issues/522), but we have not gotten MultiContainerInterfaces to work with get_class yet.

```python
# ------------------
# Create the spec
# ------------------
import pynwb
from pynwb import load_namespaces
from pynwb.spec import NWBNamespaceBuilder, NWBGroupSpec

name = 'multicontainer'
ns_path = name + ".namespace.yaml"
ext_source = name + ".extensions.yaml"

potato = NWBGroupSpec(neurodata_type_def='Potato',
                      neurodata_type_inc='NWBDataInterface',
                      doc='time of multicontainer', 
                      quantity='*'
                      )

potato_sack = NWBGroupSpec(neurodata_type_def='PotatoSack',
                           neurodata_type_inc='NWBDataInterface',
                           name='potato_sack',
                           doc='test of multicontainer', 
                           groups=[potato]
                           )

ns_builder = NWBNamespaceBuilder(name + ' extensions', name)
ns_builder.add_spec(ext_source, potato_sack)
ns_builder.export(ns_path)

# ------------------
# Load and try to use the spec
# ------------------
load_namespaces(ns_path)
Potato = get_class('Potato', 'multicontainer')
PotatoSack = get_class('PotatoSack', 'multicontainer')

my_potatos = PotatoSack(name='my_potatos', 
                        potatos=[Potato(name='One Potato'), 
                                 Potato(name='Two Potato')])

>> TypeError: incorrect type for 'potatos' (got 'list', expected 'Container')
```

Perhaps as expected, we see that PotatoSack is a NWBDataInterface, not a MultiContainerInterface. We then tried changing the neurodata_type_inc of PotatoSack to MultiContainerInterface, this also resulted in an NWBDataInterface, not a MultiContainerInterface.

- **Python Executable:** Conda
- **Python Version:** Python 3.6.7
- **Operating System:** macOS
- **Pynwb Version:** 1.0.1.post0.dev19

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?

### PR DESCRIPTION
**Pull request by bendichter**
---
name: Pull request
about: Create a pull request to help us improve
title: ''
labels: ''
assignees: ''

---

## Motivation

fix https://github.com/NeurodataWithoutBorders/pynwb/issues/888

Trying to automatically incorporate `MultiContainerInterface` but running into some issues.

## How to test the behavior?
```python
# ------------------
# Create the spec
# ------------------
from pynwb import load_namespaces, get_class
from pynwb.spec import NWBNamespaceBuilder, NWBGroupSpec

name = 'multicontainer'
ns_path = name + ".namespace.yaml"
ext_source = name + ".extensions.yaml"

potato = NWBGroupSpec(neurodata_type_def='Potato',
                      neurodata_type_inc='NWBDataInterface',
                      doc='time of multicontainer',
                      quantity='*'
                      )

potato_sack = NWBGroupSpec(neurodata_type_def='PotatoSack',
                           neurodata_type_inc='NWBDataInterface',
                           name='potato_sack',
                           doc='test of multicontainer',
                           groups=[potato]
                           )

ns_builder = NWBNamespaceBuilder(name + ' extensions', name)
ns_builder.add_spec(ext_source, potato_sack)
ns_builder.export(ns_path)

# ------------------
# Load and try to use the spec
# ------------------
load_namespaces(ns_path)
Potato = get_class('Potato', 'multicontainer')
PotatoSack = get_class('PotatoSack', 'multicontainer')

my_potatos = PotatoSack(name='my_potatos',
                        potatos=[Potato(name='One Potato'),
                                 Potato(name='Two Potato')])

print(my_potatos)
```

## Checklist

- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR description clearly describes problem and the solution?
- [x] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [x] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ? By including "Fix #XXX" you allow GitHub to close the corresponding issue.

### CODE REVIEW

File: src/hdmf/build/map.py [hunk 1]
```diff
@@ -1337,6 +1343,20 @@ def __get_cls_dict(self, base, addl_fields):
                     new_args.append(f)
                 if issubclass(dtype, (Container, Data, DataRegion)):
                     fields.append({'name': f, 'child': True})
+                    if hasattr(field_spec, 'quantity') and field_spec.quantity == '*':
```
Comment (ajtritt) [inline src/hdmf/build/map.py]: Check for quantity '+' (i.e. _one_ or more) as well

File: src/hdmf/build/map.py [hunk 2]
```diff
@@ -1337,6 +1343,20 @@ def __get_cls_dict(self, base, addl_fields):
...
+                        if field_spec.data_type_def is not None:
+                            name = field_spec.data_type_def
+                        elif field_spec.data_type_inc is not None:
+                            name = field_spec.data_type_inc
+                        else:
+                            raise ValueError('no neurodata type defined for {}'.format(f))
```
Comment (ajtritt) [inline src/hdmf/build/map.py]: Remove mention of neurodata

File: src/hdmf/build/map.py [hunk 3]
```diff
@@ -1387,6 +1410,8 @@ def get_container_cls(self, **kwargs):
                 if not spec.is_inherited_spec(field_spec):
                     fields[k] = field_spec
             d = self.__get_cls_dict(parent_cls, fields)
+            if '__clsconf__' in d:
```
Comment (ajtritt) [inline src/hdmf/build/map.py]: Do not assume `__clsconf__` indicates _MultiContainerInterface_

**Review by ajtritt — CHANGES_REQUESTED** [review]
MultiContainerInterface would need to be pulled over from pynwb, as well.

File: src/hdmf/build/map.py (hunk 1 above)
Comment (bendichter) [reply src/hdmf/build/map.py]:
```suggestion
                    if hasattr(field_spec, 'quantity') and field_spec.quantity in ('*', '+'):
```

File: src/hdmf/build/map.py (hunk 2 above)
Comment (bendichter) [reply src/hdmf/build/map.py]:
```suggestion
                            raise ValueError('no hdmf data type defined for {}'.format(f))
```

File: src/hdmf/build/map.py (hunk 3 above)
Comment (bendichter) [reply src/hdmf/build/map.py]:
```suggestion
            if any(hasattr(field_spec, 'quantity') and field_spec.quantity in ('*', '+') for field in addl_fields):
```

**Comment by bendichter** [comment]
@rly I'm not passing tests because I'm actually getting a `TypeDoesNotExistError` here:
https://github.com/hdmf-dev/hdmf/blob/4b9078e0fa8d952183aeb07c82a92aa85f0e3712/tests/unit/build_tests/test_io_map.py#L289

Which is it supposed to be?

File: src/hdmf/build/map.py [hunk 4]
```diff
@@ -1480,6 +1502,9 @@ def get_container_cls(self, **kwargs):
             for k, field_spec in attr_names.items():
                 if not spec.is_inherited_spec(field_spec):
                     fields[k] = field_spec
+            d = self.__get_cls_dict(parent_cls, fields)
```
Comment (rly) [inline src/hdmf/build/map.py]: What is the purpose of this line vs line 1509 below?

**Comment by rly** [comment]
@bendichter see my comment in map.py. The TypeDoesNotExistError should be caught and a ValueError should be thrown instead.

File: src/hdmf/build/map.py (hunk 4 above)
Comment (bendichter) [reply src/hdmf/build/map.py]: @rly good catch! That was a merge mistake. Deleting that line solves the issue.

**Comment by bendichter** [comment]
@ajtritt, @rly some of the methods of `MultiContainerInterface` rely on methods of `NWBBaseType`, including the use of `fields`. Now that I have moved `MultiContainerInterface` into `hdmf` inheriting from `Container` these methods are no longer available. What would we do here? Was there a reason why we left those methods in `pynwb` in `NWBBaseType` and didn't carry them over to `hdmf`?

**Comment by bendichter** [comment]
I guess this #60 answers my question (the answer is we should move a lot of `pynwb.NWBBaseType` to `hdmf.Container`)

**Comment by rly** [comment]
This has been done by #399.

**Pull request closed without merging**